[te] Fix bugs with shift operators

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49271 [te] Fix bugs with shift operators**
* #49247 [te] Ban uint8 tensors from fusion groups

Two things:

1. These throw exceptions in their constructor, which causes a segfault (*), so
   move the exceptions to ::make.
2. They technically support FP types but the rules are complicated so let's not
   bother.

(*) The reason for the segfault: all Exprs including these inherit from
KernelScopedObject, whose constructor adds the object to a list for destruction
at the end of the containing KernelArena's lifetime.  But if the derived-class
constructor throws, the object is deleted even though it's still in the
KernelArena's list.  So when the KernelArena is itself deleted, it double-frees
the pointer and dies.  I've also fixed And, Or, and Xor in this diff.

Differential Revision: [D25512052](https://our.internmc.facebook.com/intern/diff/D25512052/)

bertmaher committed:
[te] Fix bugs with shift operators

Two things:

1. These throw exceptions in their constructor, which causes a segfault (*), so
   move the exceptions to ::make.
2. They technically support FP types but the rules are complicated so let's not
   bother.

(*) The reason for the segfault: all Exprs including these inherit from
KernelScopedObject, whose constructor adds the object to a list for destruction
at the end of the containing KernelArena's lifetime.  But if the derived-class
constructor throws, the object is deleted even though it's still in the
KernelArena's list.  So when the KernelArena is itself deleted, it double-frees
the pointer and dies.  I've also fixed And, Or, and Xor in this diff.

Differential Revision: [D25512052](https://our.internmc.facebook.com/intern/diff/D25512052/)

[ghstack-poisoned]

Krovatkin commented on torch/csrc/jit/tensorexpr/ir.h:
```diff
@@ -150,69 +150,51 @@ class Mod : public BinaryOpNode<Mod> {
-    if (lhs->dtype() != rhs->dtype()) {
-      throw malformed_input("bad dtype in And");
+    if (lhs.dtype() != rhs.dtype()) {
+      throw malformed_input("lhs/rhs dtype mismatch");
```
out of curiosity: these throws used to be in a c-tor?

bertmaher replied:
Yeah, throwing in an `Expr` ctor is bad news because of the way `KernelScopedObject` and `KernelArena` work.

Krovatkin replied:
thanks!

Krovatkin reviewed (approved):
:shipit:

Krovatkin commented on torch/csrc/jit/tensorexpr/ir.h:
```diff
@@ -150,69 +150,51 @@ class Mod : public BinaryOpNode<Mod> {
-    if (lhs->dtype() != rhs->dtype()) {
-      throw malformed_input("bad dtype in And");
+    if (lhs.dtype() != rhs.dtype()) {
+      throw malformed_input("lhs/rhs dtype mismatch");
```
to double check, we can't do `(uint64_t)42 << (uint8_t)5 ?

bertmaher replied:
Yeah the TE dsl requires operand types to match, and when coming from pytorch we enforce that at the `kernel` translation layer.

Krovatkin replied:
thanks!

bertmaher committed:
fix test on "[te] Fix bugs with shift operators"


Two things:

1. These throw exceptions in their constructor, which causes a segfault (*), so
   move the exceptions to ::make.
2. They technically support FP types but the rules are complicated so let's not
   bother.

(*) The reason for the segfault: all Exprs including these inherit from
KernelScopedObject, whose constructor adds the object to a list for destruction
at the end of the containing KernelArena's lifetime.  But if the derived-class
constructor throws, the object is deleted even though it's still in the
KernelArena's list.  So when the KernelArena is itself deleted, it double-frees
the pointer and dies.  I've also fixed And, Or, and Xor in this diff.

Differential Revision: [D25512052](https://our.internmc.facebook.com/intern/diff/D25512052/)

[ghstack-poisoned]

penguinwu commented on torch/csrc/jit/tensorexpr/ir.h:
```diff
@@ -179,69 +179,51 @@ class Mod : public BinaryOpNode<Mod> {
-      throw malformed_input("bad dtype in Lshift");
-    }
-  }
+      : BitwiseOpNode(lhs, rhs, IRNodeType::kLshift) {}
```
Where can I find the definition of whatever operation that maps to `kLshift`? Intuitively I thought shift operators would take a tensor as 1st operand and an integer as 2nd operand (e.g., `tensorB = tensorA << 1`). Or does it really support sth like `tensorB = tensorA << tensorC`?

ZolotukhinM replied:
[native_functions.yaml](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml#L5488-L5498) is a good source - it lists all (or majority?) the ops that we have in pytorch.
Another useful list is what ops we accept from the fuser: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/tensorexpr_fuser.cpp#L137-L140
And one more useful place is what we know how to lower to TE:
https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/tensorexpr/kernel.cpp#L879

penguinwu replied:
Thanks @ZolotukhinM  `lshift` is defined in `native_functions.yaml`. And it supports the 2nd operand to be of tensor type or scalar type.

```
- func: __ilshift__.Scalar(Tensor(a!) self, Scalar other) -> Tensor(a!)
use_c10_dispatcher: full
variants: method
dispatch:
CPU, CUDA: __ilshift__
- func: __ilshift__.Tensor(Tensor(a!) self, Tensor other) -> Tensor(a!)
use_c10_dispatcher: full
variants: method
dispatch:
CPU, CUDA: __ilshift__
```

In the constructor of `kLshift`, we assume both operands to be tensor types. What happens if the 2nd operand is a scalar? Is it promoted to tensor types somewhere?

bertmaher replied:
The `Lshift` constructor actually just takes `Expr`s, which can either be a scalar or a tensor element.  For the scalar case we'd generate something like `lhs(i) << scalar`, for the tensor case we'd have `lhs(i) << rhs(i)`. (where `i` is standing in for whatever indexing expression we have for the tensor).

In building the TE from a jit::Graph (`kernel.cpp`) we have a lookup function called `tensorOrConstant` that will lookup a `jit::Value` in a map and return the appropriate `Expr`, whether it's a tensor or a constant.

bertmaher replied:
Oh, also, I don't think we handle non-constant scalars in the TE codegen.  I don't remember why but it wouldn't surprise me if there are a ton of corner cases since scalars are coming from the python type system.

bertmaher committed:
rebase on "[te] Fix bugs with shift operators"


Two things:

1. These throw exceptions in their constructor, which causes a segfault (*), so
   move the exceptions to ::make.
2. They technically support FP types but the rules are complicated so let's not
   bother.

(*) The reason for the segfault: all Exprs including these inherit from
KernelScopedObject, whose constructor adds the object to a list for destruction
at the end of the containing KernelArena's lifetime.  But if the derived-class
constructor throws, the object is deleted even though it's still in the
KernelArena's list.  So when the KernelArena is itself deleted, it double-frees
the pointer and dies.  I've also fixed And, Or, and Xor in this diff.

Differential Revision: [D25512052](https://our.internmc.facebook.com/intern/diff/D25512052/)

[ghstack-poisoned]

bertmaher committed:
rebase on "[te] Fix bugs with shift operators"


Two things:

1. These throw exceptions in their constructor, which causes a segfault (*), so
   move the exceptions to ::make.
2. They technically support FP types but the rules are complicated so let's not
   bother.

(*) The reason for the segfault: all Exprs including these inherit from
KernelScopedObject, whose constructor adds the object to a list for destruction
at the end of the containing KernelArena's lifetime.  But if the derived-class
constructor throws, the object is deleted even though it's still in the
KernelArena's list.  So when the KernelArena is itself deleted, it double-frees
the pointer and dies.  I've also fixed And, Or, and Xor in this diff.

Differential Revision: [D25512052](https://our.internmc.facebook.com/intern/diff/D25512052/)

[ghstack-poisoned]

bertmaher committed:
fix interp on "[te] Fix bugs with shift operators"


Two things:

1. These throw exceptions in their constructor, which causes a segfault (*), so
   move the exceptions to ::make.
2. They technically support FP types but the rules are complicated so let's not
   bother.

(*) The reason for the segfault: all Exprs including these inherit from
KernelScopedObject, whose constructor adds the object to a list for destruction
at the end of the containing KernelArena's lifetime.  But if the derived-class
constructor throws, the object is deleted even though it's still in the
KernelArena's list.  So when the KernelArena is itself deleted, it double-frees
the pointer and dies.  I've also fixed And, Or, and Xor in this diff.

Differential Revision: [D25512052](https://our.internmc.facebook.com/intern/diff/D25512052/)

[ghstack-poisoned]

bertmaher committed:
fix interp on "[te] Fix bugs with shift operators"


Two things:

1. These throw exceptions in their constructor, which causes a segfault (*), so
   move the exceptions to ::make.
2. They technically support FP types but the rules are complicated so let's not
   bother.

(*) The reason for the segfault: all Exprs including these inherit from
KernelScopedObject, whose constructor adds the object to a list for destruction
at the end of the containing KernelArena's lifetime.  But if the derived-class
constructor throws, the object is deleted even though it's still in the
KernelArena's list.  So when the KernelArena is itself deleted, it double-frees
the pointer and dies.  I've also fixed And, Or, and Xor in this diff.

Differential Revision: [D25512052](https://our.internmc.facebook.com/intern/diff/D25512052/)

[ghstack-poisoned]
